main/src/library/File.flix

```diff
@@ -59,4 +159,180 @@ namespace File {
         else
             readAllHelper(r, line :: acc)
 
+    ///
+    /// Returns a string containing the given file `f`.
+    ///
+    pub def read(f: String): Result[String, String] & Impure =
+        try {
+            import new java.io.File(String) as newFile;
+            import java.io.File.toPath();
+            import java.nio.file.Files:newBufferedReader(##java.nio.file.Path);
+            import java.io.BufferedReader.close();
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let reader = newBufferedReader(javaPath);
+            let result = readHelper(readAll(reader));
+            close(reader);
+            Ok(result)
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage();
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Helper function for read.
+    /// Converts a list of strings to a string with linebreaks between every string in `list`.
+    ///
+    def readHelper(list: List[String]): String & Pure =
+        List.foldLeft((x, y) -> x + "\n" + y, "", list)
+
+    ///
+    /// Returns a string containing the given file `f` with the options `opts`.
+    ///
+    pub def readWith(opts: {offset :: Int64, count :: Int32, charSet :: String}, f : String): Result[String, String] & Impure = {
+        try {
+            import new java.io.File(String) as newFile;
+            import java.nio.charset.Charset:forName(String);
+            import java.io.File.toPath();
+            import java.nio.file.Files:newBufferedReader(##java.nio.file.Path, ##java.nio.charset.Charset);
+            import java.io.BufferedReader.close();
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let reader = newBufferedReader(javaPath, forName(opts.charSet));
+            let fileRead = readHelper(readAll(reader));
+            close(reader);
+            Ok(readWithHelper(fileRead, opts.offset, opts.count))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage();
+                Err(getMessage(ex))
+        }
+    }
+
+    ///
+    /// Helper function for readWith.
+    /// Takes a String and applys the options.
+    ///
+    def readWithHelper(f: String, offset: Int64, count: Int32): String & Pure =
+        let fileLength = String.length(f);
+        match Int64.tryToInt32(offset) {
+            case None => String.dropRight(fileLength - count, f)
+            case Some(n) =>
+                let str = String.dropLeft(n, f);
+                let strLength = String.length(str);
+                String.dropRight(strLength - count, str)
+        }
+
+    ///
+    /// Returns an iterator of the given file `f`
+    ///
+    pub def readLinesIter(f: String): Iterator[Result[String, String]] & Impure =
```

Comment:
Shouldn't this return `Result[Interator[String], String]` since the iterator doesn't ever return Err? I'm not sure but I think the iterator should throw `bug!` if the iterator is empty? this iterator keeps returning null forever

Reply:
This is discussed in the ticket. The most correct signature would be `Result[Iterator[Result[String, String]], String]` but that is a mouthful